### PR TITLE
Initial notification preference implementation

### PR DESCRIPTION
**Pull request by andela-toduah**
Add Notification Preference to Users to help manage notifications.

### CODE REVIEW

File: app/controllers/preference_controller.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+class PreferenceController < ApplicationController
+  def update
+    pref_params = preference_params
+    user = User.find_by(zhishi_id: pref_params.delete(:zhishi_id))
```
Comment (andela-ooranagwa) [inline app/controllers/preference_controller.rb]: 👍

File: app/services/base_notification_service.rb [hunk 2]
```diff
@@ -11,7 +11,19 @@ def subscribers
...
     subscribers.each do |user|
-      user.send_message(resource, service: self)
+      if check_service_allowed(user)
+        user.send_message(resource, service: self)
+      end
     end
   end
+
+  def check_service_allowed(user)
+    class_name = self.class.to_s
```
Comment (andela-ooranagwa) [inline app/services/base_notification_service.rb]: 👍

**Comment by andela-ooranagwa** [comment]
Nice work

**Comment by andela-ooranagwa** [comment]
Could we have a route also for fetching user preferences?

**Comment by andela-oakinniranye** [comment]
Can this also create the default preferences and ensure all users have their preferences set up after creation.
Also since it's going to be a direct link from the FE to this and we're using the same `JWT` logic on Zhishi, you could also accept the token params to fetch the `current_user`, since FE would has access to this token.

**Comment by andela-toduah** [comment]
@andela-ooranagwa That's true. We'll factor that in.

**Comment by andela-toduah** [comment]
@andela-oakinniranye The seedfile already has code to create preferences for current users, and there's a `after_create` hook that handles creating preferences for every new user.

File: config/routes.rb [hunk 3]
```diff
@@ -3,61 +3,8 @@
...
-  #   resources :photos, concerns: :toggleable
-
-  # Example resource route within a namespace:
-  #   namespace :admin do
-  #     # Directs /admin/products/* to Admin::ProductsController
-  #     # (app/controllers/admin/products_controller.rb)
-  #     resources :products
-  #   end
+  post "ai", to: "ai#index"
+  put "user/preference/update", to: "preference#update"
```
Comment (andela-oakinniranye) [inline config/routes.rb]: could this rather be `user/:id/preference`. and use the http methods to decide the action to take

Comment (andela-toduah) [inline config/routes.rb]: True. It's been implemented.

**Comment by andela-toduah** [comment]
andela-ooranagwa A route for fetching preferences has been added.

**Comment by andela-oakinniranye** [comment]
👍

**Pull request merged**